source/adapters/hip/device.cpp

```diff
@@ -950,8 +951,57 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGetNativeHandle(
 }
 
 UR_APIEXPORT ur_result_t UR_APICALL urDeviceCreateWithNativeHandle(
-    ur_native_handle_t, ur_platform_handle_t,
-    const ur_device_native_properties_t *, ur_device_handle_t *) {
+    ur_native_handle_t hNativeDevice, ur_platform_handle_t hPlatform,
+    const ur_device_native_properties_t *pProperties,
+    ur_device_handle_t *phDevice) {
+  std::ignore = pProperties;
+
+  // We can't cast between ur_native_handle_t and hipDevice_t, so memcpy the
+  // bits instead
+  hipDevice_t HIPDevice = 0;
+  memcpy(&HIPDevice, &hNativeDevice, sizeof(hipDevice_t));
+
+  auto IsDevice = [=](std::unique_ptr<ur_device_handle_t_> &Dev) {
+    return Dev->get() == HIPDevice;
+  };
+
+  // If a platform is provided just check if the device is in it
+  if (hPlatform) {
+    auto SearchRes = std::find_if(begin(hPlatform->Devices),
+                                  end(hPlatform->Devices), IsDevice);
+    if (SearchRes != end(hPlatform->Devices)) {
+      *phDevice = SearchRes->get();
+      return UR_RESULT_SUCCESS;
+    }
+  }
+
+  // Get list of platforms
+  uint32_t NumPlatforms = 0;
+  ur_adapter_handle_t AdapterHandle = &adapter;
+  ur_result_t Result =
+      urPlatformGet(&AdapterHandle, 1, 0, nullptr, &NumPlatforms);
```

Comment:
nit:

I think with the exception of opencl adapter, every adapter can only have a max of one platform.
So I think you can just check that NumPlatforms is 1 here. If not you can return `UR_RESULT_ERROR_INVALID_OPERATION` directly at this point.
Then later you don't need to have a loop over the platforms.

Although it doesn't really make much difference. So whatever you prefer.

Reply:
Oh, thanks. So if that's the case of the CUDA adapter, I wonder why there's a loop there. Should we fix that up at some point?

Reply:
> Oh, thanks. So if that's the case of the CUDA adapter, I wonder why there's a loop there. Should we fix that up at some point?
Sure
I guess it is because it was originally copied from the opencl implementation where there are presumably always (at least) two platforms? cpu and gpu.

Reply:
The CUDA adapter has one platform per device, that only changed with the multi-device context patch, that has already been merged for the HIP adapter

Reply:
Thanks for the *context*. Will do a follow-up patch.